src/slice-chop.c

```diff
@@ -2,60 +2,144 @@
- *   case, otherwise will be `vec_size(indices)`.
- * @member out The list container for the result.
+ * Index manager/generator for chopping purposes
+ *
+ * There are 3 types of possible indices:
+ * - If `indices = NULL, sizes = NULL`, then we use a sequential size 1 index
+ *   that just increments by 1 from `0` to `vec_size(x) - 1`.
+ * - If `indices` is supplied, then each element of `indices` is an integer
+ *   vector of locations to chop with.
+ * - If `sizes` is supplied, then each element of `sizes` is the size of the
+ *   current slice to chop. The sizes are accumulated in order to get the
+ *   start location of the next slice.
+ *
+ * - Generate the next index with `indices_next()`.
+ * - Generate the output size with `indices_out_size()`.
+ *
+ * @member shelter The shelter to protect the entire chop indices manager.
+ * @member indices, v_indices
+ *   - If `NULL`, then `indices` aren't being used. `v_indices` is set to
+ *     `NULL`.
+ *   - Otherwise, a list of integer vector indices to chop with. `v_indices` is
+ *     set to `r_list_cbegin(indices)`.
+ * @member sizes, v_sizes
+ *   - If `NULL`, then `sizes` aren't being used. `v_sizes` is set to `NULL`.
+ *   - Otherwise, an integer vector of sequential sizes to chop with. `v_sizes`
+ *     is set to `r_int_cbegin(sizes)`.
+ * @member index, p_index
+ *   - If neither `indices` nor `sizes` are provided, `index` is a scalar
+ *     integer vector that starts at 0 and is incremented by 1 at every
+ *     iteration. `p_index` points to `r_int_begin(index)` and is used to
+ *     perform the increment.
+ *   - If `indices` is provided, this is set to the i-th element of `indices`
+ *     at each iteration, and `p_index` is set to `NULL`.
+ *   - If `sizes` is provided, this is a compact-seq representing the i-th
+ *     slice. `p_index` points to `r_int_begin(index)` and is used to updated
+ *     the compact-seq at each iteration.
+ * @member has_indices Whether or not `indices` was provided.
+ * @member has_sizes Whether or not `sizes` was provided.
+ * @member loc The current iteration value.
  */
-struct vctrs_chop_info {
+struct vctrs_chop_indices {
   r_obj* shelter;
 
-  struct vctrs_proxy_info proxy_info;
+  r_obj* indices;
+  r_obj* const* v_indices;
+
+  r_obj* sizes;
+  const int* v_sizes;
 
   r_obj* index;
   int* p_index;
+
   bool has_indices;
+  bool has_sizes;
 
-  r_ssize out_size;
-  r_obj* out;
+  r_ssize loc;
 };
 
 #include "decl/slice-chop-decl.h"
 
 // -----------------------------------------------------------------------------
 
 static
-struct vctrs_chop_info new_chop_info(r_obj* x, r_obj* indices) {
-  struct vctrs_chop_info info;
-  info.shelter = KEEP(r_alloc_list(3));
+struct vctrs_chop_indices* new_chop_indices(r_obj* x, r_obj* indices, r_obj* sizes) {
+  r_obj* shelter = KEEP(r_alloc_list(4));
+
+  r_obj* self = r_alloc_raw(sizeof(struct vctrs_chop_indices));
+  r_list_poke(shelter, 0, self);
```

Comment:
Also switched to our more standard approach of returning a pointer that manages itself